File: 練習問題/1-5/問題5：二分探索（バイナリサーチ）.py
import random


def binary_search(arr: list[int], target: int) -> bool:
    low: int = 0
    high: int = len(arr) - 1

    while low <= high:
        mid: int = (low + high) // 2
        if (target == arr[mid]):
            return True
        elif (target < arr[mid]):
            high = mid - 1
        else:
            low = mid + 1
    return False


def random_target() -> int:
    return random.randint(0, 100)


def random_arr(n: int) -> list[int]:
    return sorted(random.sample(range(5, 101), n))


def main() -> None:
    arr = random_arr(random_target())
    target = random_target()
    found = binary_search(arr, target)

    print(f"target: {target}")
    print(f"searchArr: {arr}")
    print(found)


main()

File: 練習問題/1-5/問題5：二分探索（バイナリサーチ）.ts
const randomTarget = (): number => {
    return Math.floor(Math.random() * 101);
};

const randomArr = (n: number): number[] => {
    const pool = Array.from({ length: 101 }, (_, i) => i);

    for (let i: number = 0; i < n; i++) {
        let j = i + Math.floor(Math.random() * (101 - i));
        [pool[i], pool[j]] = [pool[j], pool[i]];
    }

    return pool.slice(0, n).sort((a, b) => a - b);
};

const binarySearch = (arr: number[], target: number): boolean => {
    let low: number = 0;
    let high: number = arr.length - 1;

    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        if (target === arr[mid]) {
            return true;
        } else if (target < arr[mid]) {
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    return false;
};

const target = randomTarget(); // できれば 5〜100 に揃える
const arr = randomArr(randomTarget());
console.log(
    'target: ' + target,
    '\narr: [' + arr + ']',
    '\n' + binarySearch(arr, target),
);
